packages/core/core/src/TargetResolver.js

```diff
@@ -37,6 +43,362 @@ const DEFAULT_PRODUCTION_ENGINES = {
 const DEFAULT_DIST_DIRNAME = 'dist';
 const COMMON_TARGETS = ['main', 'module', 'browser', 'types'];
 
+function generateDescriptorError({
+  pkgContents,
+  pkgPath,
+  targetName,
+  keys,
+  message,
+  hints
+}: {|
+  pkgContents: ?string,
+  pkgPath: ?string,
+  targetName: string,
+  keys: Array<{|key: string, type?: ?'key' | 'value', message?: string|}>,
+  message: string,
+  hints?: ?Array<string>
+|}) {
+  let highlight;
+  if (pkgContents) {
+    highlight = generateJSONCodeHighlights(
+      pkgContents,
+      keys.map(({key, type, message}) => ({
+        key: `/targets/${targetName}/${key}`,
+        type: type,
+        message
+      }))
+    );
+  }
+  throw new ThrowableDiagnostic({
+    diagnostic: {
+      message,
+      origin: '@parcel/core',
+      filePath: pkgPath || undefined,
+      language: 'json',
+      codeFrame:
+        highlight && pkgContents
+          ? {
+              code: pkgContents,
+              codeHighlights: highlight
+            }
+          : undefined,
+      hints: hints || undefined
+    }
+  });
+}
+
+function parseEngines(
+  targetName: ?string,
+  engines: mixed
+): Engines | typeof undefined {
+  if (engines === undefined) {
+    return;
+  } else if (engines && typeof engines === 'object') {
+    let result = {};
+    for (let key in engines) {
+      let value: mixed = engines[key];
+      if (key === 'browsers' && Array.isArray(value)) {
+        value = value.map(v => {
+          if (typeof v !== 'string') {
+            throw new Error(
```

Comment:
Should these be diagnostics too?

Reply:
Correct, the push yesterday was more of a POC  😄